### DISCUSSION TITLE
Script - Reading i2c sensor values

### DISCUSSION THREAD

**Discussion by letshin** [arendst/Tasmota#17551] [Q&A] [answered]
Is it possible to use script to read i2c sensor values?

Trying to interface with: https://14core.com/wp-content/uploads/2017/12/Sensirion-Mass-Flow-Meters-SFM3000-I2C-Functional-Description.pdf but have no experience working with i2c at all beyond using libraries.

And have tried 
```
>S
val1=ia(0x40)
if val1>0{
iw1(0x80 0x1000)
val3=ir1(0x81)
}
```
With the logic that the datasheet says that measurements start when the start measurement command (0x1000) is sent to the sensor. I don't know where to go from here beyond a vague idea that information is read from 0x80. Can anyone point me in the right direction with some general instructions/milestones of what needs to be done please so I can try that.

**Comment by gemu2015** [comment, marked as answer]
the flowmeter needs to be started only once, try this
```
>D
found=0
val=0
>B
; check device
found=ia(0x40)
if found>0 {
; start measurement, write cmd hex 1000 
iw2(0xffff 0x1000)
} else {
print device not found
}
>F
if found>0 {
; read word (2 bytes)
val=ir2(0xffff)
; raw value should be scaled to give physical value
print %val%
}
```

**Reply by letshin** [reply]
Thanks. I tried this and am getting the max readout from 0x81 but would have been expecting 32000 for zero flow.
```
10:19:25.070 65535.00
```
Trying to understand this and was going through the datasheet again. Is ACK/NACK handled by the script library already? I was rereading the datasheet and for reading the 0x81 address there were a few scenarios and am not sure if this needs to be coded in manually. 
```
81(NACK) FF(ACK) FF(ACK) FF(NACK) - First measurement result read out after chip reset is invalid
81(ACK) F0(ACK) 00(ACK) CRC(NACK) - Valid measurement data is read out + CRC byte.
81(ACK) F0(ACK) 14(NACK) - Updated flow value is read out. CRC is not read out.
81(NACK) - No valid measuremen
```
To read the CRC byte, should I shift 2 bytes and read a single byte off 0x81?

**Comment by gemu2015** [comment]
ok this device has no registers at all, so you need to omit the register transfer. i modified the source to allow this.
if register address is 0xffff then no register address is transferred.
try the modified script above.

**Reply by letshin** [reply]
I am getting this error whilst compiling- `tasmota/tasmota_xdrv_driver/xdrv_10_scripter.ino:2585:20: error: conversion from 'IPAddress' to 'float' is ambiguous`. I think its new code since I can't find it in the older version of xdrv_10_scripter.ino. This is one of two errors I'm seeing so I'm not sure if this is stopping it from compiling.

I'm not sure if the other error is related to you as well since it has your name in it haha:
```
Compiling .pio\build\tasmota\lib429\Epaper_29-gemu-1.0\epd2in9.cpp.o
*** [.pio\build\tasmota\src\tasmota.ino.cpp.o] Error 1
```

I also missed swapping the a for (ACKs) in the `81(NACK) FF(ACK) FF(ACK) FF(NACK)` example I posted earlier. I hope this doesn't complicate things.

BTW, what do you mean by no registers? The datasheet says that the 0x1000 command "Starts continuous measurement and/or move internal pointer (back) to flow measurement result register". How do I know in the future the difference between these devices? Is it because of the FFFF value?
![image](https://user-images.githubusercontent.com/40326519/210185953-4babb339-7d86-4371-8ed3-b5a9792caaab.png)


Building was attempted on vscode 1.74.2 and the latest Tasmota development branch on github, swapping xdrv_10_scripter.ino for your current version in xdrv_10_scripter.ino.zip.

**Reply by gemu2015** [reply]
this will compile.

you should however not see this epaper error???

"normal" i2c devices have addresses within the device which you must send first before accessing a device register.
this device does not have that. you shall not send a register select byte.

i developed a driver for this device 6 years ago at work (medical application, not for arduino).
i have several of these at work, but since i retired i am only occasionally there. otherwise i could test immediately with tasmota.

**Reply by letshin** [reply]
Compiled it, it went through. Not sure where the epaper error came from previously but no problem compiling now.
The updated script is now reading 
```
20:05:50.495 1367.00
```
Blowing into the flowmeter induces no change in the value being read out. As a sanity check, I had previously tried the same wiring/setup with the Sensirion Arduino library and can confirm that it works there.

I don't know if this helps but the Arduino Sensirion library has the following to send the 0x1000 command, only partially pasted here (https://github.com/Sensirion/arduino-i2c-sfm3000/tree/main/src and https://github.com/Sensirion/arduino-core/tree/main/src):
```
uint16_t
SensirionI2CSfm3000::startContinuousMeasurement() {
    uint16_t error;
    uint8_t buffer[2];
    SensirionI2CTxFrame txFrame =
        SensirionI2CTxFrame::createWithUInt16Command(0x1000, buffer, 2,
                                                     CRC31_00);

    error =
        SensirionI2CCommunication::sendFrame(_i2cAddress, txFrame, *_i2cBus);
    delay(1);
    return error;
}

SensirionI2CTxFrame SensirionI2CTxFrame::createWithUInt16Command(
    uint16_t command, uint8_t buffer[], size_t bufferSize, CrcPolynomial poly) {
    SensirionI2CTxFrame instance =
        SensirionI2CTxFrame(buffer, bufferSize, 2, poly);
    instance._buffer[0] = static_cast<uint8_t>((command & 0xFF00) >> 8);
    instance._buffer[1] = static_cast<uint8_t>((command & 0x00FF) >> 0);
    return instance;
}
```
The data read command (also only partially pasted here) is:
```
uint16_t SensirionI2CSfm3000::readMeasurementRaw(uint16_t& flowRaw) {
    uint16_t error;
    uint8_t buffer[3];

    delay(1);

    SensirionI2CRxFrame rxFrame(buffer, 3);
    error = SensirionI2CCommunication::receiveFrame(_i2cAddress, 3, rxFrame,
                                                    *_i2cBus,
                                                    CRC31_00);
    if (error) {
        return error;
    }

    error |= rxFrame.getUInt16(flowRaw);
    return error;
}

uint16_t SensirionRxFrame::getUInt16(uint16_t& data) {
    if (_numBytes < 2) {
        return RxFrameError | NoDataError;
    }
    data = static_cast<uint16_t>(_buffer[_index++]) << 8;
    data |= static_cast<uint16_t>(_buffer[_index++]);
    _numBytes -= 2;
    return NoError;
}

```

**Reply by gemu2015** [reply]
ok, bytes are send low byte first, but must be sent high byte first 
so try
; start measurement, write cmd hex 1000 
iw2(0xffff 0x0010)

**Reply by letshin** [reply]
Yes! This now works. Thank you so much. How did you tell its high byte first? Is it from:
```
    instance._buffer[0] = static_cast<uint8_t>((command & 0xFF00) >> 8);
    instance._buffer[1] = static_cast<uint8_t>((command & 0x00FF) >> 0);
```
Would you be planning to add this functionality to Tasmota script as well?

""normal" i2c devices have addresses within the device which you must send first before accessing a device register."
Do you mean like the MCP3021 where I can get data by `iw2 (0x4d)`. Is iw2 equivalent to Arduino's `Wire.requestFrom(0x4d, 2u)`? 
```
// SYNTAX
Wire.requestFrom(address, quantity);
Wire.requestFrom(address, quantity, stop);
```
https://docs.particle.io/reference/device-os/api/wire-i2c/isenabled/

**Reply by gemu2015** [reply]
it should now work as expected with
iw2(0xffff 0x1000)

i now have made the byte order selectable.

if you confirm this working i will make a pr to tasmota development.

> Do you mean like the MCP3021 where I can get data by iw2 (0x4d). Is iw2 equivalent to Arduino's Wire.requestFrom(0x4d, 2u)?

yes that is the equivalent

**Reply by letshin** [reply]
Had compile errors:
```
/tasmota_backups/Tasmota-development/tasmota/tasmota_xdrv_driver/xdrv_10_scripter.ino:128:17: note: #pragma message: script compression option used
 #pragma message "script compression option used"
                 ^
/tasmota_backups/Tasmota-development/tasmota/tasmota_xdrv_driver/xdrv_10_scripter.ino: In function 'uint32_t script_i2c(uint8_t, uint16_t, uint32_t)':

/tasmota_backups/Tasmota-development/tasmota/tasmota_xdrv_driver/xdrv_10_scripter.ino:10674:9: error: expected ',' or ';' before 'for'
       cert = (char*)malloc(fsize +2);
         ^
/tasmota_backups/Tasmota-development/tasmota/tasmota_xdrv_driver/xdrv_10_scripter.ino:10674:31: error: 'cnt' was not declared in this scope       
       cert = (char*)malloc(fsize +2);
                               ^
/tasmota_backups/Tasmota-development/tasmota/tasmota_xdrv_driver/xdrv_10_scripter.ino:10674:49: error: expected ';' before ')' token
       cert = (char*)malloc(fsize +2);
```

**Reply by gemu2015** [reply]
sorry try this one
[xdrv_10_scripter.ino.zip](https://github.com/arendst/Tasmota/files/10337242/xdrv_10_scripter.ino.zip)

**Reply by letshin** [reply]
Thanks. this works now:
```
>D
found=0
val=0
val1=0
>B
; check device
found=ia(0x40)
if found>0 {
; start measurement, write cmd hex 1000 
iw2(0xffff 0x1000)
print SUCCESS send 0x1000 to 0x80
} else {
print FAIL send 0x1000 to 0x80
}
>S
if found>0 {
; read word (2 bytes)
val=ir2(0xffff)
val1=(val-32000)/140
; raw value should be scaled to give physical value
print %val%
print %val1%
}
```
Results:

![image](https://user-images.githubusercontent.com/40326519/210386612-111e419c-06a2-4f41-94e6-2b720e03d88b.png)

If you can spare some time, could you please help explain why we had to choose to send 0x1000 to 0xffff and why the datasheet mentioned 0x80/0x81 and what its use is when its not required? I'm interested in understanding (at least at surface level how to read future datasheets) what's happening here. Thanks again.

**Comment by gemu2015** [comment]
normally i2c devices have device address and several registers with addresses normally in the range of 0 to 255.
the device address of this device is 0x80 but must be given as 0x40 because it is shifted by 1  bit 0 is read/write bit

since this device does not have any internal register addresses the address shall not be sent.

to tell the driver that the address must be omitted we set the address to 0xffff. this is out of the normal address range and is actually nothing but a flag.
 
the change in scripter has been merged into development now.
